core/src/main/java/ai/timefold/solver/core/impl/domain/entity/descriptor/EntityForEachFilter.java

```diff
@@ -0,0 +1,86 @@
+package ai.timefold.solver.core.impl.domain.entity.descriptor;
+
+import java.util.Collection;
+import java.util.function.Predicate;
+
+import ai.timefold.solver.core.impl.domain.variable.declarative.DeclarativeShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.ShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.VariableDescriptor;
+import ai.timefold.solver.core.preview.api.domain.variable.declarative.ShadowVariablesInconsistent;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public class EntityForEachFilter {
+    private final Predicate<Object> assignedAndConsistentPredicate;
+
+    // null when there are no declarative shadow variables,
+    // since a for each node with a null filter is treated as an always true filter
+    @Nullable
+    private final Predicate<Object> consistentPredicate;
+
+    EntityForEachFilter(EntityDescriptor<?> entityDescriptor) {
+        var solutionDescriptor = entityDescriptor.getSolutionDescriptor();
+        var listVariableDescriptor = solutionDescriptor.getListVariableDescriptor();
+        var declarativeShadowVariables = getDeclarativeShadowVariables(entityDescriptor);
+        var shadowVariableInconsistentDescriptor = entityDescriptor.getShadowVariablesInconsistentDescriptor();
+
+        if (!declarativeShadowVariables.isEmpty() && shadowVariableInconsistentDescriptor == null) {
+            // TODO: Remove this when supplier is present
+            throw new IllegalStateException(
+                    "Entity class %s has declarative shadow variables (%s) but does not have a @%s member."
```

Comment:
Do we want to keep calling it "declarative shadow variables" if there are no other ones?

Even in the docs, we'll change it. The word "declarative" will go away - it's just shadow variables.